mdp/utils/covariance.py

```diff
@@ -133,6 +135,146 @@ def fix(self, center=True):
         return cov_mtx, avg, tlen
 
 
+class VartimeCovarianceMatrix(CovarianceMatrix):
+    """This class stores an empirical covariance matrix that can be updated
+    incrementally. A call to the 'fix' method returns the current state of
+    the covariance matrix, the average and the number of observations, and
+    resets the internal data.
+
+    As compared to the *CovarianceMatrix* class, this class accepts unevenly
+    sampled input in conjunction with a time increment between samples. The
+    covariance matrix is then computed as a (centered) scalar product
+    between functions, that is sampled unevenly, using the trapezoid rule.
+
+    The mean is computed using the trapezoid rule as well.
+
+    Note that the internal sum is a standard __add__ operation.
+    """
+
+    def __init__(self, dtype=None):
+        """Initialize an *VartimeCovarianceMatrix*.
+
+        :param dtype: Datatype of the input.
+            Default is None. If dtype is not defined, it will be inherited
+            from the first data bunch received by 'update'. 
+        :type dtype: numpy.dtype or str
+        """
+        super(VartimeCovarianceMatrix,
+              self).__init__(dtype=dtype, bias=True)
+        # we need to keep steps and actual time
+        # values seperately, due to the weighing
+        # which is not required in the evenly sampled case
+        self._steps = 0
+        self.tphase = 0
+
+    def update(self, x, dt=None):
+        """Update internal structures.
+
+        Note that no consistency checks are performed on the data (this is
+        typically done in the enclosing node).
+
+        :param x: Timed sequence of vectors, with samples along the rows
+            and random variables along the columns.
+        :type x: numpy.ndarray
+
+        :param dt: Sequence of time increments between vectors. Must be
+            of length *x.shape[0]-1* in case of a single call. In case of
+            multiple calls with intended time dependence - as opposed to
+            computing the results as a weighted avarage of separate calls -
+            *dt* must be of length *x.shape[0]* starting with the second call.
+            The first element will be considered as the time difference
+            between the last element of *x* in the preceding call and the
+            first element of *x* in the current call. In case of not supplying
+            a *x.shape[0]*-th element in *dt*, time dependence will be dropped.
+            If *dt* omitted entirely, it will be considered to be one everywhere
+            whereas time dependence between calls is still dropped.
```

Comment:
Like mentioned earlier, better write this in cases-style. That said, there should be an explicit way to control whether or not inter-chunk derivative is computed. Can we have an additional arg to control this behavior? E.g.
`continuous == True`: Compute derivative to previous chunk (default)
`continuous == False`: Dismiss derivative to previous chunk

I think even in the first call, if `continuous == True` there would be an interpretation for `dt[0]` if `len(dt) == x.shape[0]`: for weightinh the fist sample w.r.t. to trapezoid rule.
I have a bit mixed feelings about how to deal with inconsistent values for `continuous` and `len(dt)`.
For sake of usability I think I'd prefer it lie this:
`continuous==True and len(dt) == x.shape[0]`: Okay. If in first call, ignore `dt[0]` or use it for weighting.
`continuous==True and len(dt) == x.shape[0]-1`: Raise appropriate exception.
`continuous==False and len(dt) == x.shape[0]`: Ignore `dt[0]`. Raise warning unless in first call.
`continuous==False and len(dt) == x.shape[0]-1`: Okay.
Other length of `dt`:  Raise appropriate exception.
If `dt` is a number, i.e. has no length: Use this number as constant time increment. Consider time derivative to previous chunk according to value of `continuous`.
`dt is None`: Like previous case with `dt=1.` (default)

Reply:
Please forward this behavior to the train method of VartimeSFA.

Reply:
I vote against another parameter, as then - like you describe - dt and the parameter need to be matched to have a valid input. That is why originally I supplied the "special" dt's separately. That way it is an explicit argument to a certain use case and nothing needs to be matched. I still prefer the current implementation over an extra argument.

Reply:
That would mean that length of dt is the only way to control time dependence between chunks. That's a rather subtle way and may easily lead to unexpected or inaccurate behavior. Maybe the effect is small if chunks are large, but quantifying the effect, especially given unknown usecases is another story/beyond the goal here.
Also, it uglifies the user's code: Instead of setting a flag if the user wants to change this mode of operation, slicing is necessary - harder to read, less expressive. Imagine someone reads code using this feature: Without looking into the MDP doc carefully -- really in low level detail --, it would be hard to understand what chunks etc means. An explicit argument is a much clearer pointer that there is this mode of operation to consider. Slicing details of dt in user code is a weak eye-opener that there may be more semantics to it. Making this mode of operation explicit is the safe way.

Secondly: Imagine someone needs to use this node with constant time inc (e.g. because non-constant time incs only occur in a minority of the chunks). With current usage there is no way to have time dependence between chunks while using e.g. `dt=1.2` rather than dt being an array. Yes, you can set `dt=numx.ones(whatever)*1.2` but that complicates usage and is less performant - consider large chunks, this wastes memory too.

Lastly, adding another arg is not harmful because it would have a default value. I guess most usually, time dependence is the reasonable behavior so it should be the default. That's another issue with the current version: E.g. `dt=1.2` behaves against this default.